### DISCUSSION TITLE
Android 12 built-in IKEv2 issues

### DISCUSSION THREAD

**Discussion by Unb0rn** [strongswan/strongswan#1220] [Q&A]
Hey! I wanted to try internal Android 12 IKEv2 client, but it looks like there are some problems (while strongSwan android client works fine):
```
conn ikev2-psk-android
  keyexchange	 = ikev2
  fragmentation = yes
  mobike           = yes

  rightauth	= eap-mschapv2
  eap_identity  = %identity

  rightsourceip = 10.10.40.0/24

  rightsendcert = never
```
Journal log on connect and android logcat with stacktrace are attached:
[android-ikev2.log](https://github.com/strongswan/strongswan/files/9316751/android-ikev2.log)
[strongswan.log](https://github.com/strongswan/strongswan/files/9316753/strongswan.log)

This stacktrace line makes me feel like this is somehow related to algo negotiation:
`android.net.ipsec.ike.exceptions.AuthenticationFailedException: Unrecognized ASN.1 objects for Signature algorithm and Hash`

Running strongSwan on Ubuntu 22.04. `Linux strongSwan U5.9.5/K5.15.0-46-generic`

**Comment by Thermi** [comment]
If that is indeed the case then I guess they do not like the "identity" signature hash algorithm. Because that one's the only different one.

**Comment by tobiasbrunner** [comment]
It could be an issue with the `identity` hash algorithm, as they definitely don't support it (5 is the identifier for the "Identity" hash algorithm from [RFC 8420](https://datatracker.ietf.org/doc/html/rfc8420)):

    IkeAuthDigitalSignPayload: Unexpected or repeated Signature Hash Algorithm: 5

But seems weird to not just ignore that and select a known hash algorithm and signature scheme.

So it could also be an issue with the certificate/key you configured for the client. Maybe it uses a key type that the client doesn't support. Please provide more information about that.

**Reply by Unb0rn** [reply]
Well, maybe I don't get the question)
In fact I have two connections set up, the first one is:
```
conn ikev2-eap
  keyexchange	 = ikev2
  fragmentation = yes
  mobike           = yes

  leftid        = "CN=myserver.example.com"
  leftcert      = myserver.example.com.crt

  rightauth     = eap-mschapv2

  rightsourceip = 10.10.50.0/24

  rightsendcert = never
  eap_identity  = %identity
```
It uses ecc-based certificate (secp384r1) and works great with strongSwan android client. But this connection is ignored if I try to connect with native Android 12 client. By trial and error I've found that the problem disappears if I remove those `leftid` and `lefcert` statements.
Should I add them back and  replace ecc-based cert with rsa one? I thought that Android does not verify the server at all, just ignores any certs sent its way...
Apart from this both connections are the same and right side (road warrior) auth works as eap-mschapv2 password for both cases (in fact, the user and password are the same)

**Reply by tobiasbrunner** [reply]
There is no server certificate yet when IKE_SA_INIT is processed. So if you don't actually have a certificate configured on the client, which it might be loading/checking at that point, it could really just be an issue with the "identity" hash algorithm. Maybe it tries to figure out what potential OIDs could correspond to specific hash algorithms (even though it doesn't support the alg and doesn't even need any of those because the client is not authenticated with a certificate).

It is theoretically possible to configure specific authentication schemes, which will result in a reduced list of hash algorithms. But that configuration is only available for `pubkey` authentication not `eap-mschapv2` (we technically don't need those hashes for the latter, but since we need to return the notify to enable the signature auth extension, in order to use it to authenticate the server, we still need to return the supported hash algos).

> By trial and error I've found that the problem disappears if I remove those leftid and lefcert statements.

The client probably sends a specific remote identity that doesn't match your certificate's subject DN/`leftid`. But since you need a server certificate loaded, the connection won't really be usable without `leftcert`.

**Reply by Unb0rn** [reply]
It looks like you're right!
Changing subdomain to one with the rsa-based cert (in ipsec.secrets) seems to change the Android stacktrace to:
`android.net.ipsec.ike.exceptions.AuthenticationFailedException: The remote/server failed to provide a end certificate`
So setting `leftsendcert = always` without manually specifying the cert or id seems to finally solve this issue. Maybe we should note it somewhere or add some workarounds?

So, to summarize, only RSA server cert is supported (specifying ECC one will bring this ASN.1 error)
The server should not specify the cert or id, it should be auto-selected from the domain name (or the connection will be ignored)
The server should send the cert always without CR payload...
Thanks!

**Reply by tobiasbrunner** [reply]
> So setting leftsendcert = always without manually specifying the cert or id seems to finally solve this issue.

Without `leftcert` no certificate is loaded, unless there is one loaded with another connection. And without `leftid` (and `leftcert`) the identity will default to the IP address, which then has to be contained as subjectAltName in a loaded certificate for which a private key is available. All that you could also configure explicitly to make clear what's actually expected to happen.

> So, to summarize, only RSA server cert is supported (specifying ECC one will bring this ASN.1 error)

OK, but that hasn't really anything to do with the config. It just means the client might not support ECDSA certs (or are you using EdDSA certs?). If that's the case, just load the certificate with `leftcert` and configure an appropriate identity that's contained in the certificate (and depending on what the client is using/sending as IDr).

> it should be auto-selected from the domain name

That's not something that's actually happening.

>  The server should send the cert always without CR payload...

Yes, `leftsendcert=always` is necessary with some clients (but could also depend on their config).

**Comment by doriantaylor** [comment]
Just doing a drive-by here to mention that the client doesn't seem to like traffic selectors either. Backtrace:

```
10-05 17:19:46.962  1731 12411 D IkeV2VpnRunner: ChildOpened for network 101
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner: Error in ChildOpened for network 101
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner: java.lang.IllegalArgumentException: Invalid range; start address must be before end address
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at com.android.net.module.util.IpRange.<init>(IpRange.java:60)
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at com.android.server.connectivity.VpnIkev2Utils.getRoutesFromTrafficSelectors(VpnIkev2Utils.java:449)
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at com.android.server.connectivity.Vpn$IkeV2VpnRunner.onChildOpened(Vpn.java:2551)
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at com.android.server.connectivity.VpnIkev2Utils$ChildSessionCallbackImpl.onOpened(VpnIkev2Utils.java:345)
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at com.android.internal.net.ipsec.ike.ChildSessionStateMachine$CreateChildLocalCreateBase.lambda$validateAndBuildChild$0$ChildSessionStateMachine$CreateChildLocalCreateBase(ChildSessionStateMachine.java:837)
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at com.android.internal.net.ipsec.ike.ChildSessionStateMachine$CreateChildLocalCreateBase$$ExternalSyntheticLambda0.run(Unknown Source:6)
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
10-05 17:19:46.965  1731 12411 D IkeV2VpnRunner:        at java.lang.Thread.run(Thread.java:920)
10-05 17:19:46.965  1731 12411 D Vpn     : setting state=FAILED, reason=Invalid range; start address must be before end address
```

[Here's the code where the error is.](https://cs.android.com/android/platform/superproject/+/master:frameworks/libs/net/common/framework/com/android/net/module/util/IpRange.java;l=60;drc=232175a504067d3e07bf770c868828f38ea6b7de)

Set the traffic selectors in the responder to `0.0.0.0/0` and it works. I thought it was the virtual IP behaviour (a /32 traffic selector would have start `>=` end) but it doesn't seem to care.

**Reply by tobiasbrunner** [reply]
> Set the traffic selectors in the responder to `0.0.0.0/0` and it works.

What were the traffic selectors you tried before? Did you use strongSwan as server? (Note that all traffic selectors are sent as ranges, even `0.0.0.0/0`, that's simply `0.0.0.0 - 255.255.255.255`).

> I thought it was the virtual IP behaviour (a /32 traffic selector would have start `>=` end) but it doesn't seem to care.

It probably only calls this for remote traffic selectors, since the goal of `getRoutesFromTrafficSelectors()` is presumably to get a list of routes to remote subnets to install.

**Reply by doriantaylor** [reply]
You're right. I have a TS for the /32 of the actual machine it's connecting to. If I change that to /31 it works. Cute.

**Reply by tobiasbrunner** [reply]
Yeah, that `>=` is definitely a mistake. The second constructor that takes an `IpPrefix` accepts a `/32` or `/128` address (there is even [a test for it](https://cs.android.com/android/platform/superproject/+/master:frameworks/libs/net/common/tests/unit/src/com/android/net/module/util/IpRangeTest.java;l=99;bpv=0;bpt=0)), so the first constructor probably should too.
